go/parquet/file/page_writer.go

```diff
@@ -0,0 +1,466 @@
+func (pw *serializedPageWriter) NumValues() int64             { return pw.nvalues }
+func (pw *serializedPageWriter) DictionaryPageOffset() int64  { return pw.dictPageOffset }
+func (pw *serializedPageWriter) DataPageoffset() int64        { return pw.dataPageOffset }
+func (pw *serializedPageWriter) TotalCompressedSize() int64   { return pw.totalCompressed }
+func (pw *serializedPageWriter) TotalUncompressedSize() int64 { return pw.totalUncompressed }
+
+func (pw *serializedPageWriter) WriteDictionaryPage(page *DictionaryPage) (int64, error) {
+	uncompressed := len(page.Data())
+
+	var data []byte
+	if pw.HasCompressor() {
+		var buffer bytes.Buffer
+		data = pw.Compress(&buffer, page.Data())
+		// data = buffer.Bytes()
+	} else {
+		data = page.Data()
+	}
+
+	dictPageHeader := &format.DictionaryPageHeader{
+		NumValues: page.NumValues(),
+		Encoding:  page.Encoding(),
+		IsSorted:  libthrift.BoolPtr(page.IsSorted()),
+	}
+
+	if pw.dataEncryptor != nil {
+		pw.updateEncryption(encryption.DictPageModule)
+		pw.encryptionBuf.Reset()
+		pw.encryptionBuf.Grow(pw.dataEncryptor.CiphertextSizeDelta() + len(data))
+		pw.dataEncryptor.Encrypt(&pw.encryptionBuf, data)
+		data = pw.encryptionBuf.Bytes()
+	}
+
+	pageHdr := pageHeaderPool.Get().(*format.PageHeader)
+	defer pageHeaderPool.Put(pageHdr)
+	pageHdr.Type = format.PageType_DICTIONARY_PAGE
+	pageHdr.UncompressedPageSize = int32(uncompressed)
+	pageHdr.CompressedPageSize = int32(len(data))
+	pageHdr.DictionaryPageHeader = dictPageHeader
+	pageHdr.DataPageHeader = nil
+	pageHdr.DataPageHeaderV2 = nil
+
+	startPos := pw.sink.Tell()
+	if pw.dictPageOffset == 0 {
+		pw.dictPageOffset = int64(startPos)
+	}
+
+	if pw.metaEncryptor != nil {
+		if err := pw.updateEncryption(encryption.DictPageHeaderModule); err != nil {
+			return 0, err
+		}
+	}
+	headerSize, err := pw.thriftSerializer.Serialize(pageHdr, pw.sink, pw.metaEncryptor)
+	if err != nil {
+		return 0, err
+	}
+	written, err := pw.sink.Write(data)
+	if err != nil {
+		return 0, err
+	}
+
+	written += headerSize
+
+	pw.totalUncompressed += int64(uncompressed + headerSize)
+	pw.totalCompressed = int64(written)
+	pw.dictEncodingStats[parquet.Encoding(page.encoding)]++
+	return int64(written), nil
+}
+
+var pageHeaderPool = sync.Pool{
+	New: func() interface{} {
+		return format.NewPageHeader()
+	},
+}
+
+func (pw *serializedPageWriter) WriteDataPage(page DataPage) (int64, error) {
+	uncompressed := page.UncompressedSize()
+	data := page.Data()
+
+	if pw.dataEncryptor != nil {
+		if err := pw.updateEncryption(encryption.DataPageModule); err != nil {
+			return 0, err
+		}
+		pw.encryptionBuf.Reset()
+		pw.encryptionBuf.Grow(pw.dataEncryptor.CiphertextSizeDelta() + len(data))
+		pw.dataEncryptor.Encrypt(&pw.encryptionBuf, data)
+		data = pw.encryptionBuf.Bytes()
+	}
+
+	pageHdr := pageHeaderPool.Get().(*format.PageHeader)
+	defer pageHeaderPool.Put(pageHdr)
+	pageHdr.UncompressedPageSize = int32(uncompressed)
```

Comment:
is overflow possible here?

Reply:
so i just double checked the `parquet.thrift` file and realized that the `uncompressed_size` for data pages is an int32, not an int64. So i've updated the `DataPage` structs accordingly to be int32 for the sizes of the pages the whole way through, making sure that we avoid overflows since we'll always read an int32 from the file and the writer should enforce the max page size in the options.